lektor/watcher.py

```diff
@@ -1,60 +1,103 @@
 import os
 import queue
 import time
+from collections import OrderedDict
+from itertools import zip_longest
 
+import click
 from watchdog.events import DirModifiedEvent
 from watchdog.events import FileMovedEvent
 from watchdog.events import FileSystemEventHandler
 from watchdog.observers import Observer
+from watchdog.observers.polling import PollingObserver
 
 from lektor.utils import get_cache_dir
 
-# Alias this as this can be called during interpreter shutdown
-_Empty = queue.Empty
-
 
 class EventHandler(FileSystemEventHandler):
-    def __init__(self, callback=None):
-        if callback is not None:
-            self.queue = None
-            self.callback = callback
-        else:
-            self.queue = queue.Queue()
-            self.callback = self.queue.put
+    def __init__(self):
+        self.queue = queue.Queue()
 
     def on_any_event(self, event):
         if not isinstance(event, DirModifiedEvent):
-            path = (
-                event.dest_path if isinstance(event, FileMovedEvent) else event.src_path
-            )
-            item = (time.time(), event.event_type, path)
-            if self.queue is not None:
-                self.queue.put(item)
+            if isinstance(event, FileMovedEvent):
+                path = event.dest_path
             else:
-                self.callback(*item)
+                path = event.src_path
+            self.queue.put((time.time(), event.event_type, path))
+
+
+def _fullname(cls):
+    """Return the full name of a class (including the module name)."""
+    return f"{cls.__module__}.{cls.__qualname__}"
+
+
+def _unique_everseen(seq):
+    """Return the unique elements in sequence, preserving order."""
+    return OrderedDict.fromkeys(seq).keys()
 
 
 class BasicWatcher:
-    def __init__(self, paths, callback=None):
-        self.event_handler = EventHandler(callback=callback)
-        self.observer = Observer()
-        for path in paths:
-            self.observer.schedule(self.event_handler, path, recursive=True)
-        self.observer.setDaemon(True)
+    def __init__(self, paths, observer_classes=(Observer, PollingObserver)):
+        self.event_handler = EventHandler()
+        self.paths = paths
+        self.observer_classes = observer_classes
+        self.observer = None
+
+    def start(self):
+        observer_classes = list(_unique_everseen(self.observer_classes))
```

Comment:
Could you put a comment here to explain why you need to remove duplicates? I'd expect a function that I pass a sequence of classes to try to try these classes in sequence - if I pass something in twice, I'd expect it to be tried twice.

Alternatively this could also be made explicit by conditionally defining the default observers to avoid a duplicate PollingObserver.